okhttp/src/main/java/okhttp3/internal/Util.kt

```diff
@@ -0,0 +1,642 @@
+      throw AssertionError("Invalid IPv6 address: '$host'")
+    }
+
+    try {
+      val result = IDN.toASCII(host).toLowerCase(Locale.US)
+      if (result.isEmpty()) return null
+
+      // Confirm that the IDN ToASCII result doesn't contain any illegal characters.
+      return if (containsInvalidHostnameAsciiCodes(result)) {
+        null
+      } else result
+      // TODO: implement all label limits.
+    } catch (e: IllegalArgumentException) {
+      return null
+    }
+  }
+
+  private fun containsInvalidHostnameAsciiCodes(hostnameAscii: String): Boolean {
+    for (i in 0 until hostnameAscii.length) {
+      val c = hostnameAscii[i]
+      // The WHATWG Host parsing rules accepts some character codes which are invalid by
+      // definition for OkHttp's host header checks (and the WHATWG Host syntax definition). Here
+      // we rule out characters that would cause problems in host headers.
+      if (c <= '\u001f' || c >= '\u007f') {
+        return true
+      }
+      // Check for the characters mentioned in the WHATWG Host parsing spec:
+      // U+0000, U+0009, U+000A, U+000D, U+0020, "#", "%", "/", ":", "?", "@", "[", "\", and "]"
+      // (excluding the characters covered above).
+      if (" #%/:?@[\\]".indexOf(c) != -1) {
+        return true
+      }
+    }
+    return false
+  }
+
+  /**
+   * Returns the index of the first character in [input] that is either a control character
+   * (like `\u0000` or `\n`) or a non-ASCII character. Returns -1 if [input] has no such
+   * characters.
+   */
+  fun indexOfControlOrNonAscii(input: String): Int {
+    var i = 0
+    val length = input.length
+    while (i < length) {
+      val c = input[i]
+      if (c <= '\u001f' || c >= '\u007f') {
+        return i
+      }
+      i++
+    }
+    return -1
+  }
+
+  /** Returns true if [host] is not a host name and might be an IP address.  */
+  @JvmStatic
+  fun verifyAsIpAddress(host: String): Boolean {
+    return VERIFY_AS_IP_ADDRESS.matches(host)
+  }
+
+  /** Returns a [Locale.US] formatted [String].  */
+  @JvmStatic
+  fun format(format: String, vararg args: Any): String {
+    return String.format(Locale.US, format, *args)
+  }
+
+  @Throws(IOException::class)
+  fun bomAwareCharset(source: BufferedSource, charset: Charset): Charset {
+    return when (source.select(UNICODE_BOMS)) {
+      0 -> UTF_8
+      1 -> UTF_16BE
+      2 -> UTF_16LE
+      3 -> UTF_32BE
+      4 -> UTF_32LE
+      -1 -> charset
+      else -> throw AssertionError()
+    }
+  }
+
+  fun checkDuration(name: String, duration: Long, unit: TimeUnit?): Int {
+    if (duration < 0) throw IllegalArgumentException("$name < 0")
+    if (unit == null) throw NullPointerException("unit == null")
+    val millis = unit.toMillis(duration)
+    if (millis > Integer.MAX_VALUE) throw IllegalArgumentException("$name too large.")
+    if (millis == 0L && duration > 0) throw IllegalArgumentException("$name too small.")
```

Comment:
require

Reply:
There are subtle changes here in exception types, I made the first two checks, and second two require